src/main/java/org/icgc/argo/program_service/services/InvitationService.java

```diff
@@ -45,38 +51,71 @@ public UUID inviteUser(@NotNull ProgramEntity program,
     @NotNull UserRole role) {
+    val programShortName = program.getShortName();
     val invitation = new JoinProgramInviteEntity(program, userEmail, firstName, lastName, role);
+    val previousInvitations = invitationRepository.findAllByProgramShortNameAndUserEmail(programShortName, userEmail);
+
+    // 1) If there is an accepted invitation, then return an error saying the user has accepted the invitation
+    if (previousInvitations.stream().anyMatch(i -> i.getStatus() == ACCEPTED)) {
+      throw Status.ALREADY_EXISTS.augmentDescription("Invitation was already accepted").asRuntimeException();
+    }
+    // 2) If there is a PENDING invitation, then that previous invitation should be set to INVALID.
+    //If a previous invitation is EXPIRED,REVOKED or INVALID, leave them as is.
+    previousInvitations.stream().filter(i -> i.getStatus() == PENDING).map(i -> i.setStatus(INVALID)).
+      forEach(i -> invitationRepository.save(i));
+
     invitationRepository.save(invitation);
     mailService.sendInviteEmail(invitation);
     return invitation.getId();
   }
 
-  @Transactional
-  public List<JoinProgramInviteEntity> listInvitations(@NonNull String programShortName) {
-    return invitationRepository.findAllByProgramShortName(programShortName);
-  }
-
   @Transactional
   public EgoUser acceptInvite(@NonNull UUID invitationId) throws NotFoundException {
     val invitation = invitationRepository
       .findById(invitationId)
       .orElseThrow(() ->
         new NotFoundException(format("Cannot find invitation with id '%s' ", invitationId)));
+
+    if (invitation.getStatus() != PENDING) {
+      if (invitation.getStatus() == null) {
+        throw Status.FAILED_PRECONDITION.augmentDescription(
+          "Cannot accept invitation because it is in state(NULL), not PENDING").asRuntimeException();
+      }
+      throw Status.FAILED_PRECONDITION.augmentDescription(
+        format("Cannot accept invitation because it is in state(%s), not PENDING", invitation.getStatus().toString())).
+        asRuntimeException();
+    }
+
     invitation.accept();
     invitationRepository.save(invitation);
     egoService.joinProgram(invitation.getUserEmail(), invitation.getProgram().getShortName(), invitation.getRole());
     return egoService.convertInvitationToEgoUser(invitation);
   }
 
-  public Optional<JoinProgramInviteEntity> getInvitation(String programShortName, String email) {
-    return invitationRepository.findTopByProgramShortNameAndUserEmailOrderByCreatedAtDesc(programShortName, email);
+  public Optional<JoinProgramInviteEntity> getLatestInvitation(String programShortName, String email) {
+    val invitations = listInvitations(programShortName, email);
+    val validInvitations = invitations.stream().filter(i -> i.getStatus() != INVALID && i.getStatus() != REVOKED).
+      collect(Collectors.toList());
+    if (validInvitations.size() == 0) {
+      return Optional.empty();
+    }
+    return Optional.of(validInvitations.get(0));
```

Comment:
Does `get(0)` return the latest invitation?

Reply:
In this case, yes. It gets the first element in the list, which should have been sorted in descending order by creation date.